contracts/zero-ex/contracts/src/transformers/FillQuoteTransformer.sol

```diff
@@ -388,31 +387,108 @@ contract FillQuoteTransformer is
+                    (order, takerAssetFillAmount, signature)
+                returns (IExchange.FillResults memory fillResults)
+            {
+                // Update maker quantity based on changes in token balances.
+                results.makerTokenBoughtAmount = makerToken.balanceOf(address(this))
+                    .safeSub(initialMakerTokenBalance);
+                // We can trust the other fill result quantities.
+                results.protocolFeePaid = fillResults.protocolFeePaid;
+                results.takerTokenSoldAmount = fillResults.takerAssetFilledAmount;
+                // If the taker fee is payable in the taker asset, include the
+                // taker fee in the total amount sold.
+                if (isTakerFeeInTakerToken) {
+                    results.takerTokenSoldAmount =
+                        results.takerTokenSoldAmount.safeAdd(fillResults.takerFeePaid);
+                }
+            } catch (bytes memory) {
+                // Swallow failures, leaving all results as zero.
+            }
         }
     }
 
+    /// @dev Attempt to fill an ERC20 Bridge order. If the fill reverts,
+    ///      or the amount filled was not sufficient this reverts.
+    /// @param order The bridge order to fill.
+    /// @param inputTokenAmount How much taker asset to fill.
+    /// @param outputTokenAmount How much maker asset to receive.
+    /// @param makerToken The maker token.
+    function fillBridgeOrder(
+        IExchange.Order calldata order,
+        uint256 inputTokenAmount,
+        uint256 outputTokenAmount,
+        IERC20TokenV06 makerToken
+    )
+        external
+        returns (FillOrderResults memory results)
+    {
+        // Track changes in the maker token balance.
+        uint256 initialMakerTokenBalance = makerToken.balanceOf(address(this));
+        (
+            address tokenAddress,
+            address bridgeAddress,
+            bytes memory bridgeData
+        ) = abi.decode(
+            order.makerAssetData.sliceDestructive(4, order.makerAssetData.length),
+            (address, address, bytes)
+        );
+        require(bridgeAddress != address(this), "INVALID_BRIDGE_ADDRESS");
+        // Transfer the tokens to the bridge to perform the work
+        _getTokenFromERC20AssetData(order.takerAssetData).compatTransfer(
+            bridgeAddress,
+            inputTokenAmount
+        );
+        IERC20Bridge(bridgeAddress).bridgeTransferFrom(
+            tokenAddress,
+            order.makerAddress,
+            address(this),
+            outputTokenAmount, // amount to transfer back from the bridge
+            bridgeData
+        );
+        uint256 afterMakerTokenBalance = makerToken.balanceOf(address(this));
+        // Ensure that the maker token balance has increased by the expected amount
+        require(
+            afterMakerTokenBalance >= initialMakerTokenBalance.safeAdd(outputTokenAmount),
+            "BRIDGE_UNDERPAY"
+        );
```

Comment:
What do we think of removing this check since we have the important one at the end anyway?

Reply:
We only check that we sold X amount or bought Y amount. This checks that we sold to the bridge at a reasonably expected price. If we didn't revert we could've sold 100 for 1 instead of 100 for 99. This needs to stay for safe bridge market sells.

Reply:
Won't the `transformERC2()` check assert the price?
https://github.com/0xProject/0x-monorepo/blob/development/contracts/zero-ex/contracts/src/features/TransformERC20.sol#L285

Reply:
Yeah looks like it 😂